presto-main/src/main/java/io/prestosql/server/security/oauth2/OAuth2Service.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.server.security.oauth2;
+
+import com.github.scribejava.core.model.OAuth2AccessToken;
+import com.github.scribejava.core.model.OAuth2AccessTokenErrorResponse;
+import com.github.scribejava.core.model.OAuthConstants;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableMap;
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jws;
+import io.jsonwebtoken.JwtParser;
+import io.jsonwebtoken.Jwts;
+import io.prestosql.server.security.oauth2.Challenge.Started;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.prestosql.server.security.oauth2.OAuth2Resource.CALLBACK_ENDPOINT;
+import static io.prestosql.server.security.oauth2.OAuth2Resource.OAUTH2_API_PREFIX;
+import static java.util.Objects.requireNonNull;
+
+public class OAuth2Service
```

Comment:
I'm not sure if I agree. In my opinion having the logic of a stateful challenge encapsulated in a single class improves the readability. However, I'll try to improve it even further but without inlining since storing challenges in a cache is going to be needed (see: https://github.com/prestosql/presto/pull/5355#discussion_r517507229).